File: backend-dotnetWebApi/Services/ImportState.cs
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace backend_dotnetWebApi.Services
{
    public class ImportState
    {
        private readonly object _lock = new object();
        public string Status { get; set; } = "idle"; // idle | running | done | error
        public int Total { get; set; } = 0;
        public int Current { get; set; } = 0;
        private List<string> _log = new List<string>();

        public void Log(string message)
        {
            lock (_lock)
            {
                _log.Add(message);
                if (_log.Count > 200)
                {
                    _log = _log.GetRange(_log.Count - 200, 200);
                }
            }
        }

        public object GetStateSnapshot()
        {
            lock (_lock)
            {
                return new
                {
                    status = Status,
                    total = Total,
                    current = Current,
                    log = new List<string>(_log)
                };
            }
        }

        public void Reset()
        {
            lock (_lock)
            {
                Status = "running";
                Total = 0;
                Current = 0;
                _log.Clear();
            }
        }
    }
}

File: backend-springboot/src/main/java/com/anime/backend/service/ImportState.java
package com.anime.backend.service;

import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Component
public class ImportState {

    private String status = "idle"; // idle | running | done | error
    private int total = 0;
    private int current = 0;
    private List<String> log = new ArrayList<>();

    public synchronized void logMessage(String message) {
        log.add(message);
        if (log.size() > 200) {
            log = new ArrayList<>(log.subList(log.size() - 200, log.size()));
        }
    }

    public synchronized Map<String, Object> getStateSnapshot() {
        return Map.of(
            "status", status,
            "total", total,
            "current", current,
            "log", new ArrayList<>(log)
        );
    }

    public synchronized void reset() {
        this.status = "running";
        this.total = 0;
        this.current = 0;
        this.log.clear();
    }

    public synchronized void setStatus(String status) {
        this.status = status;
    }

    public synchronized void setTotal(int total) {
        this.total = total;
    }

    public synchronized void setCurrent(int current) {
        this.current = current;
    }
    
    public synchronized int getTotal() {
        return this.total;
    }
}
